### PR TITLE
Chore/update to plug 3.0

### PR DESCRIPTION
**Pull request by aliXsed**
This is the first brick in creating Cennznet 2.0. Next stage would be "enabling smart contracts". However we need to merge this with develop first.

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -89,9 +89,6 @@ zeroize = { opt-level = 3 }
 
 [features]
 default = []
-quick_staking = [
```
Comment (jordy25519) [inline Cargo.toml]: keep this please, it's for building a client with quick staking sessions/eras

File: cli/Cargo.toml [hunk 2]
```diff
@@ -15,77 +15,75 @@ futures = { version = "0.3.1", features = ["compat"] }
...
 [build-dependencies]
-substrate-build-script-utils = { rev = "2.0.5", git = "https://github.com/plugblockchain/plug-blockchain" }
+substrate-build-script-utils = { rev = "v3.0.0-rc4", git = "https://github.com/plugblockchain/plug-blockchain" }
 
 [features]
 default = []
-# Uses integration test config for quicker staking eras + epochs
-quick_staking = [
-	"cennznet-runtime/integration_config"
-]
```
Comment (jordy25519) [inline cli/Cargo.toml]: keep pls

File: crml/cennzx/src/mock.rs [hunk 3]
```diff
@@ -25,87 +25,99 @@ pub const TRADE_ASSET_B_ID: AssetId = 3;
...
 	type OnKilledAccount = ();
 	type SystemWeightInfo = ();
+	type SS58Prefix = ();
 }
 
-impl prml_generic_asset::Trait for Test {
+parameter_types! {
+		pub const TreasuryModuleId: ModuleId = ModuleId(*b"py/trsry");
+}
+pub struct TransferImbalanceToTreasury;
```
Comment (jordy25519) [inline crml/cennzx/src/mock.rs:98]: could you add a comment on why this is implemented/used for this test runtime

File: crml/transaction-payment/rpc/runtime-api/src/lib.rs [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+//! Runtime API definition for transaction payment module.
+
+#![cfg_attr(not(feature = "std"), no_std)]
+
+use codec::Codec;
+use sp_runtime::traits::MaybeDisplay;
+
+pub use crml_transaction_payment::{FeeDetails, InclusionFee, RuntimeDispatchInfo};
+
+// TODO Fix conflicting implementations of trait `parity_scale_codec::Decode` for type `pallet_transaction_payment::RuntimeDispatchInfo<_>`:
```
Comment (jordy25519) [inline crml/transaction-payment/rpc/runtime-api/src/lib.rs]: will you open an issue or address in this PR? I think these changes should be added to `transaction-payment/src/types.rs`

File: runtime/src/lib.rs [hunk 5]
```diff
@@ -126,28 +137,66 @@ pub fn native_version() -> NativeVersion {
...
+	fn on_unbalanceds<B>(mut fees_then_tips: impl Iterator<Item = NegativeImbalance>) {
+		if let Some(fees) = fees_then_tips.next() {
+			Rewards::note_transaction_fees(fees.peek());
+			if let Some(tips) = fees_then_tips.next() {
+				Rewards::note_transaction_fees(tips.peek());
+			}
+		}
+	}
+}
+
```
Comment (jordy25519) [inline runtime/src/lib.rs]: should go to impls.rs

File: runtime/src/lib.rs [hunk 6]
```diff
@@ -126,28 +137,66 @@ pub fn native_version() -> NativeVersion {
...
 // Configure modules to include in the runtime.
 
-const AVERAGE_ON_INITIALIZE_WEIGHT: Perbill = Perbill::from_percent(10);
+/// We assume that ~10% of the block weight is consumed by `on_initialize` handlers.
+/// This is used to limit the maximal weight of a single extrinsic.
+const AVERAGE_ON_INITIALIZE_RATIO: Perbill = Perbill::from_percent(10);
+/// We allow `Normal` extrinsics to fill up the block up to 75%, the rest can be used
+/// by  Operational  extrinsics.
+const NORMAL_DISPATCH_RATIO: Perbill = Perbill::from_percent(75);
+/// We allow for 2 seconds of compute with a 6 second average block time.
```
Comment (jordy25519) [inline runtime/src/lib.rs]:
```suggestion
/// We allow for 2 seconds of compute with a 5 second average block time.
```

File: primitives/src/types.rs [hunk 7]
```diff
@@ -64,12 +64,17 @@ pub type Timestamp = u64;
...
+
+/// Block ID.
+pub type BlockId = generic::BlockId<Block>;
+
 /// The outer `FeeExchange` type. It is versioned to provide flexibility for future iterations
 /// while maintaining backward compatibility.
 #[derive(PartialEq, Eq, Clone, Encode, Decode, Debug)]
 pub enum FeeExchange<AssetId, Balance> {
 	/// A V1 FeeExchange
-	#[codec(compact)]
```
Comment (jordy25519) [inline primitives/src/types.rs:72]: why remove this?

File: crml/staking/src/lib.rs [hunk 8]
```diff
@@ -1901,7 +1901,8 @@ decl_module! {
 		/// # </weight>
 		#[weight = T::WeightInfo::reap_stash(1_u32)]
 		fn reap_stash(_origin, stash: T::AccountId) {
-			ensure!(T::Currency::total_balance(&stash).is_zero(), Error::<T>::FundedTarget);
+			let at_minimum = T::Currency::total_balance(&stash) == T::Currency::minimum_balance();
+			ensure!(at_minimum, Error::<T>::FundedTarget);
```
Comment (jordy25519) [inline crml/staking/src/lib.rs]: shouldn't this be
```suggestion
			let at_minimum = T::Currency::total_balance(&stash) <= T::Currency::minimum_balance();
			ensure!(at_minimum, Error::<T>::FundedTarget);
```

File: crml/staking/src/rewards/mod.rs [hunk 9]
```diff
@@ -569,81 +563,86 @@ mod tests {
...
+		type PalletInfo = PalletInfo;
+		type AccountData = prml_generic_asset::AccountData<AssetId>;
 		type OnNewAccount = ();
 		type OnKilledAccount = ();
 		type SystemWeightInfo = ();
+		type SS58Prefix = ();
 	}
 
-	impl prml_generic_asset::Trait for TestRuntime {
+	pub struct TransferImbalanceToTreasury;
```
Comment (jordy25519) [inline crml/staking/src/rewards/mod.rs:615]: could you add a comment on why this is implemented/used for this test runtime

**Review by jordy25519 — COMMENTED** [review]
Generally lgtm.

This should not merge to `develop` until after NFT launch, recommend using another trunk for that in the meantime.

Some things to check before a '2.0' branch gets deployed live would be (besides smart contracts):
- Review the bounties  + tips config makes sense for mainnet
- Fix decoding historic tx fee amounts

File: crml/transaction-payment/rpc/runtime-api/src/lib.rs (hunk 4 above)
Comment (aliXsed) [reply crml/transaction-payment/rpc/runtime-api/src/lib.rs]: We have this related issue open here: https://github.com/plugblockchain/plug-blockchain/issues/180. I will check if your suggested solution there would address the issue here.

File: primitives/src/types.rs (hunk 7 above)
Comment (aliXsed) [reply primitives/src/types.rs:72]: It is removed as part of fixing cennzx build issues. I need to put it back and chack what was the actual build issue.

File: Cargo.toml (hunk 1 above)
Comment (aliXsed) [reply Cargo.toml:92]: I don't find such a feature in the recent code.

Comment (aliXsed) [reply Cargo.toml]: Sure, During merge it was somehow recognised as old code.

File: cli/Cargo.toml (hunk 2 above)
Comment (aliXsed) [reply cli/Cargo.toml]: Will do

File: crml/staking/src/lib.rs (hunk 8 above)
Comment (aliXsed) [reply crml/staking/src/lib.rs]: This code is the same as Plug (and Substrate) and possibly written according to the recent behaviour of the Balances module. Balances wouldn't allow the balance to go below the minimum. For us and with generic asset, I'm not sure it's still the case. So probably safer to check "<=".

File: primitives/src/types.rs (hunk 7 above)
Comment (aliXsed) [reply primitives/src/types.rs:72]: `error: Invalid attribute on variant, only `#[codec(skip)]` and `#[codec(index = $u8)]` are accepted.`

File: crml/transaction-payment/rpc/runtime-api/src/lib.rs (hunk 4 above)
Comment (aliXsed) [reply crml/transaction-payment/rpc/runtime-api/src/lib.rs]: Fixed it according to the suggestion.

Comment (jordy25519) [reply crml/transaction-payment/rpc/runtime-api/src/lib.rs]: can clean up this file now

File: primitives/src/types.rs (hunk 7 above)
Comment (jordy25519) [reply primitives/src/types.rs:72]: Ok I think it's safe to ignore, seems like `enum` doesn't support 'compact' in codec 2.0. (or rather it's already compact)

File: crml/transaction-payment/rpc/runtime-api/src/lib.rs (hunk 4 above)
Comment (aliXsed) [reply crml/transaction-payment/rpc/runtime-api/src/lib.rs]: The clean up is done as well.

**Pull request closed without merging**